### PR TITLE
[Feature](bangc-ops): Add new binary op mluOpDCNBackwardData.

### PR DESCRIPTION
**Pull request by mahxn0**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Add new binary-op mluOpDCNBackwardData.

## 2. Modification
bangc-ops/kernels/dcn_backward_data/dcn_backward_data.cpp
bangc-ops/mlu_op.h
bangc-ops/test/mlu_op_gtest/pb_gtest/src/pb_test_tools.cpp
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_data/dcn_backward_data.cpp
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/dcn_backward_data/dcn_backward_data.h
docs/bangc-docs/BANGC-OPS-OpList.md

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [x] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [x] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [ ] BLOCK
  - [x] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test (e.g. float32/int8)
- [x] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [x] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
[MLU Hardware Time      ]: 353 (us)
[MLU Interface Time     ]: 403.358 (us)
[MLU IO Efficiency      ]: 0.0598806
[MLU Compute Efficiency ]: 0.00218556
[MLU Workspace Size     ]: -1 (Bytes)
[MLU TheoryOps          ]: 790020 (Ops)
[MLU TheoryIOs          ]: 6.49355e+06 (Bytes)
[MLU ComputeForce       ]: 1.024e+12 (op/s)
[MLU IoBandWidth        ]: 307.2 (GB/s)
[GPU Hardware Time      ]: 8.0384e-05 (us)
[GPU IO Efficiency      ]: 0.235106
[GPU Compute Efficiency ]: 0.00121334
[GPU Workspace Size     ]: 0 (Bytes)
[Diffs]:
[output1]
DIFF1: 5.732006e-08
DIFF2: 7.739434e-08
[output2]
DIFF1: 1.952590e-07
DIFF2: 2.119860e-07
[output3]
DIFF1: 1.972027e-07
DIFF2: 2.122958e-07
[^      OK ] /data/mlu-ops/backup/release_temp/dcn_backward_data/dcn_backward_data_8_manual_float32_float32_1655989916393.pb
[       OK ] dcn_backward_data/TestSuite.mluOp/99 (398 ms)
[----------] 100 tests from dcn_backward_data/TestSuite (72132 ms total)

[----------] Global test environment tear-down
[ SUMMARY  ] Total 100 cases of 1 op(s).
ALL PASSED.
[==========] 100 test cases from 1 test suite ran. (76218 ms total)
[  PASSED  ] 100 test cases.

```

Platform：MLU590
```bash
[MLU Hardware Time      ]: 63 (us)
[MLU Interface Time     ]: 134.173 (us)
[MLU IO Efficiency      ]: 0.0503282
[MLU Compute Efficiency ]: 0.00544271
[MLU Workspace Size     ]: -1 (Bytes)
[MLU TheoryOps          ]: 790020 (Ops)
[MLU TheoryIOs          ]: 6.49355e+06 (Bytes)
[MLU ComputeForce       ]: 2.304e+12 (op/s)
[MLU IoBandWidth        ]: 2048 (GB/s)
[GPU Hardware Time      ]: 8.0384e-05 (us)
[GPU IO Efficiency      ]: 0.235106
[GPU Compute Efficiency ]: 0.00121334
[GPU Workspace Size     ]: 0 (Bytes)
[Diffs]:
[output1]
DIFF1: 7.173844e-08
DIFF2: 9.022284e-08
[output2]
DIFF1: 1.952590e-07
DIFF2: 2.119860e-07
[output3]
DIFF1: 1.972027e-07
DIFF2: 2.122958e-07
[^      OK ] /data/mlu-ops/backup/release_temp/dcn_backward_data/dcn_backward_data_8_manual_float32_float32_1655989916393.pb
[       OK ] dcn_backward_data/TestSuite.mluOp/99 (39 ms)
[----------] 100 tests from dcn_backward_data/TestSuite (24846 ms total)

[----------] Global test environment tear-down
[ SUMMARY  ] Total 100 cases of 1 op(s).
ALL PASSED.
[==========] 100 test cases from 1 test suite ran. (28360 ms total)
[  PASSED  ] 100 test cases.
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

**Review by DanieeelLiu — APPROVED** [review]

File: bangc-ops/kernels/dcn_backward_data/dcn_backward_data.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include <math.h>
+#include <limits.h>
```
Comment (chqy99) [inline bangc-ops/kernels/dcn_backward_data/dcn_backward_data.cpp:24]: 字母序

**Review by chqy99 — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ */
+mluOpStatus_t MLUOP_WIN_API
+mluOpDestroyDCNDescriptor(mluOpDCNDescriptor_t dcn_desc);
+
+// Group:DCN
+/*!
+ * @brief Returns in \p workspace_size the size of the MLU memory that is used as an extra
+ * workspace to optimize the deformable convolution backward data operation.
+ *
+ * The size of extra workspace is based on the given information of the deformable convolution
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:16775]:
```suggestion
 * The size of the extra workspace is based on the information of the deformable convolution
```

File: bangc-ops/kernels/dcn_backward_data/dcn_backward_data.cpp (hunk 1 above)
Comment (mahxn0) [reply bangc-ops/kernels/dcn_backward_data/dcn_backward_data.cpp:23]: 已修改

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ * @param[in] dcn_desc
+ *   Input. The deformable convolution descriptor to be destroyed.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
+ *
+ * @note
+ * - You need to call this function after calling the ::mluOpDCNForward, ::mluOpDCNBackwardData,
+ *   or ::mluOpDCNBackwardWeight function. Otherwise, \p MLUOP_STATUS_BAD_PARAM is returned.
+ * - This function should be called to destroy the deformable convolution descriptor.
+ *   Otherwise, the memory leak may occur.
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ *
+ * @param[in] dcn_desc
+ *   Input. The deformable convolution descriptor to be destroyed.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
+ *
+ * @note
+ * - You need to call this function after calling the ::mluOpDCNForward, ::mluOpDCNBackwardData,
+ *   or ::mluOpDCNBackwardWeight function. Otherwise, \p MLUOP_STATUS_BAD_PARAM is returned.
+ * - This function should be called to destroy the deformable convolution descriptor.
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ * the ::mluOpCreateDCNDescriptor function.
+ *
+ * @param[in] dcn_desc
+ *   Input. The deformable convolution descriptor to be destroyed.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
+ *
+ * @note
+ * - You need to call this function after calling the ::mluOpDCNForward, ::mluOpDCNBackwardData,
+ *   or ::mluOpDCNBackwardWeight function. Otherwise, \p MLUOP_STATUS_BAD_PARAM is returned.
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ * @brief Destroys a deformable convolution descriptor \p dcn_desc that is previously created with
+ * the ::mluOpCreateDCNDescriptor function.
+ *
+ * @param[in] dcn_desc
+ *   Input. The deformable convolution descriptor to be destroyed.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
+ *
+ * @note
+ * - You need to call this function after calling the ::mluOpDCNForward, ::mluOpDCNBackwardData,
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ *
+ * @par Example
+ * - None.
+ */
+mluOpStatus_t MLUOP_WIN_API
+mluOpCreateDCNDescriptor(mluOpDCNDescriptor_t *dcn_desc);
+
+// Group:DCN
+/*!
+ * @brief Initializes the deformable convolution descriptor \p dcn_desc that is previously
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ *   Output. A host pointer to the deformable convolution descriptor that holds information about the
+ *   deformable convolution operation.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_NOT_INITIALIZED
+ *
+ * @par API Dependency
+ * - After calling this function, you can call the ::mluOpSetDCNDescriptor function to initialize
+ *   and set the information to the deformable convolution descriptor.
+ * - You need to call the ::mluOpDestroyDCNDescriptor function to destroy the descriptor.
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ *
+ * @param[out] dcn_desc
+ *   Output. A host pointer to the deformable convolution descriptor that holds information about the
+ *   deformable convolution operation.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_NOT_INITIALIZED
+ *
+ * @par API Dependency
+ * - After calling this function, you can call the ::mluOpSetDCNDescriptor function to initialize
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
 
+// Group:DCN
+/*!
+ * @brief Creates a descriptor pointed by \p dcn_desc for a deformable convolution forward
+ * or backward operation, and allocates memory for holding the information about the
+ * deformable convolution operation. The information is defined in ::mluOpDCNDescriptor_t.
+ * For more information about descriptor, see "Cambricon MLUOP User Guide".
+ *
+ * @param[out] dcn_desc
+ *   Output. A host pointer to the deformable convolution descriptor that holds information about the
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
                               void *workspace,
                               size_t workspace_size);
 
+// Group:DCN
+/*!
+ * @brief Creates a descriptor pointed by \p dcn_desc for a deformable convolution forward
+ * or backward operation, and allocates memory for holding the information about the
+ * deformable convolution operation. The information is defined in ::mluOpDCNDescriptor_t.
+ * For more information about descriptor, see "Cambricon MLUOP User Guide".
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

**Review by PetrelYy — APPROVED** [review]

**Review by AquamarineSu — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ * @param[in] stride
+ *   Input. An array that stores the filter stride for each dimension of the input tensor
+ *   used in the deformable convolution operation. For each dimension, the filter stride represents
+ *   the number of elements to slide over the input tensor. For 2D deformable
+ *   convolution, the stride should be set in height and width order.
+ * @param[in] dilation
+ *   Input. An array that stores the dilation factor for each dimension of the filter tensor
+ *   used in the deformable convolution operation. For each dimension, the dilation factor
+ *   represents the spacing between the kernel points. For 2D deformable convolution,
+ *   the dilation should be set in height and width order.
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ * @param[in] deformable_group
+ *   Input. The number of deformable offset groups that split the input offset along the channel
+ *   of input tensor. Each deformable group is deformed separately for detecting different input parts.
+ * @param[in] conv_group
+ *   Input. The number of groups that the input data is split by the number of channels
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -16613,6 +16615,409 @@ mluOpQuantizeBatchMatMulBCast(mluOpHandle_t handle,
...
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par API Dependency
+ * - Before calling this function, ::mluOpCreateDCNDescriptor should be called.
+ *
+ * @note
+ * - Currently, only supports 4D input tensor for deformable convolution forward
+ *   and backward operation.
+ * - The values of \p pad should be greater than or equal to 0.
```
[the remark this answers is not in the document]
Comment (mahxn0) [reply bangc-ops/mlu_op.h]: 已修改

**Pull request merged**